consensus/src/consensus/inner/mod.rs

```diff
@@ -0,0 +1,192 @@
+
+mod agent;
+mod commit;
+mod transaction;
+
+pub struct ConsensusInner {
+    pub public: Arc<Consensus>,
+    pub ledger: DynLedger,
+    pub memory_pool: MemoryPool,
+    pub storage: Arc<dyn Storage>,
+}
+
+struct LedgerData {
+    ledger: DynLedger,
+    commitments: Vec<Digest>,
+    serial_numbers: Vec<Digest>,
+    memos: Vec<Digest>,
+    ledger_digests: Vec<Digest>,
+}
+
+impl ConsensusInner {
+    /// scans uncommitted blocks for forks
+    async fn scan_forks(&mut self) -> Result<()> {
+        let blocks = self
+            .storage
+            .get_block_hashes(None, snarkos_storage::BlockFilter::NonCanonOnly)
+            .await?;
+        info!("scanning {} blocks for forks", blocks.len());
+        for hash in blocks {
+            let header = self.storage.get_block_header(&hash).await?;
+            let parent_state = self.storage.get_block_state(&header.previous_block_hash).await?;
+            match parent_state {
+                BlockStatus::Unknown => continue, // orphan
+                BlockStatus::Committed(_) => (),  // uncomitted child of canon, could be a fork or a hanging block
+                BlockStatus::Uncommitted => {
+                    // mid-fork or orphan chain, wait for head of fork
+                    continue;
+                }
+            };
+            let block = self.storage.get_block(&hash).await?;
+
+            self.try_commit_block(&hash, &block).await?;
+        }
+        Ok(())
+    }
+
+    fn fresh_ledger(&self, blocks: Vec<SerialBlock>) -> Result<LedgerData> {
+        let mut ledger = self.ledger.clone();
+        ledger.clear();
+        let mut new_commitments = vec![];
+        let mut new_serial_numbers = vec![];
+        let mut new_memos = vec![];
+        let mut new_digests = vec![];
+        for (i, block) in blocks.into_iter().enumerate() {
+            trace!("ledger recreation: processing block {}", i);
+            let mut commitments = vec![];
+            let mut serial_numbers = vec![];
+            let mut memos = vec![];
+            for transaction in block.transactions.iter() {
+                commitments.extend_from_slice(&transaction.new_commitments[..]);
+                serial_numbers.extend_from_slice(&transaction.old_serial_numbers[..]);
+                memos.push(transaction.memorandum.clone());
+            }
+            let digest = ledger.extend(&commitments[..], &serial_numbers[..], &memos[..])?;
+            new_commitments.extend(commitments);
+            new_serial_numbers.extend(serial_numbers);
+            new_memos.extend(memos);
+            new_digests.push(digest);
+        }
+        Ok(LedgerData {
+            ledger,
+            commitments: new_commitments,
+            serial_numbers: new_serial_numbers,
+            memos: new_memos,
+            ledger_digests: new_digests,
+        })
+    }
+
+    async fn diff_canon(&self) -> Result<()> {
+        let blocks = self.storage.get_canon_blocks(Some(128)).await?;
```

Comment:
What's the rationale behind this number? Also, I presume this function is for debug purposes? It could use a doc describing its functionality.